File: cairnkit/intent.py
"""IntentGate — route a request onto a path mode (M2).

A deterministic heuristic the orchestrator can use as a default; the model is free to
override the suggestion with ``state set-path-mode``. Keeping a testable Python heuristic
means small requests do not silently run the full 16-stage pipeline.
"""

from __future__ import annotations

from dataclasses import dataclass

# Signals that a request is a tiny single-point change.
_SINGLE_HINTS = (
    "typo", "rename", "bump", "comment", "log message", "one line", "one-line",
    "tweak", "constant", "config value", "version",
)
# Signals there is no UI/frontend surface (backend-only → lite path).
_FRONTEND_HINTS = (
    "ui", "page", "screen", "component", "css", "frontend", "front-end", "visual",
    "button", "layout", "style",
)


@dataclass(frozen=True)
class IntentResult:
    path_mode: str
    reason: str


def classify(text: str) -> IntentResult:
    """Suggest full | lite | single from the request text (heuristic, overridable)."""
    low = text.lower()
    words = len(low.split())

    if words <= 12 and any(h in low for h in _SINGLE_HINTS):
        return IntentResult("single", "short single-point change (keyword + brevity)")
    if not any(h in low for h in _FRONTEND_HINTS):
        return IntentResult("lite", "no frontend/UI surface detected — backend-only path")
    return IntentResult("full", "frontend surface present — full pipeline")

File: src/intent.rs
//! IntentGate fallback default for headless/scripted use.
//!
//! Classification is fuzzy work, so the **orchestrator (the model) is the primary
//! classifier** — it reads the request and picks `full`/`lite`/`single` by judgment in
//! any language. This Rust heuristic is only a deterministic fallback default for headless
//! or scripted runs where no model is in the loop. When unsure it defaults to `full`, so it
//! never under-processes a request.

const SINGLE_HINTS: &[&str] = &[
    "typo",
    "rename",
    "bump",
    "comment",
    "log message",
    "one line",
    "one-line",
    "tweak",
    "constant",
    "config value",
    "version",
];

pub struct IntentResult {
    pub path_mode: &'static str,
    pub reason: &'static str,
}

pub fn classify(text: &str) -> IntentResult {
    let low = text.to_lowercase();
    let words = low.split_whitespace().count();

    if words <= 12 && SINGLE_HINTS.iter().any(|h| low.contains(h)) {
        return IntentResult {
            path_mode: "single",
            reason: "short single-point change (keyword + brevity)",
        };
    }
    IntentResult {
        path_mode: "full",
        reason: "default to full pipeline; the orchestrator refines to lite/single by judgment",
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn routes() {
        // Short single-point change → single.
        assert_eq!(classify("fix a typo in the README").path_mode, "single");
        // The heuristic no longer guesses lite — backend and frontend both default to full.
        assert_eq!(
            classify("add a coupon endpoint to the order service with rate limiting").path_mode,
            "full"
        );
        assert_eq!(
            classify("build a new checkout page UI component with a styled button").path_mode,
            "full"
        );
    }
}
